### ISSUE TITLE
How to calculate the localization performance?

### ISSUE THREAD

**Issue by huaxinxiao** [jia-wan/GeneralizedLoss-Counting-Pytorch#3]
It seems that the test.py only output the density map. But how to calculate the localization performance? A hard threshold or other strategy?

**Comment by jia-wan** [comment]
Hi huaxinxiao. Thanks for your interest. Given a density map as the input we directly compute the local maximum as the candidates and a threshold is used to filter out low-density points. Then, the performance is computed according to the original NWPU or UCF papers. See below code for details:

prob_outputs = F.interpolate(density_map, scale_factor=8, mode='bilinear')
maxpool_output = F.max_pool2d(prob_outputs, 3,1,1)
maxpool_output = torch.eq(maxpool_output, prob_outputs)
maxpool_output = maxpool_output.type(torch.cuda.FloatTensor) * prob_outputs
maxpool_output = maxpool_output[0][0].detach().cpu().numpy()
maxpool_output[maxpool_output < th] = 0
y, x = maxpool_output.nonzero()

**Comment by knightyxp** [comment]
hey jia wan:
      I have a lot of questions about the above test code of the location. First of all, in the training process, we input Keypoints, which are the marked points. I tried the method you mentioned above, but the count is always 0. The input density map you mentioned above uses the threshold to filter low confidence points. However, to locate the points this is not realistic for your method because your method input is an image and the output is keypoints, not a density map. I get confused about the location evaluation of GL Loss /DM Count ,  and I am urgent to solve this problem. looking forward to your reply. my Gmail is knightyxp@gmail.com

**Comment by pietz** [comment]
Not one of the authors, but the method above works for me and the results look good. I'm not sure I understand your problem entirely but the Geometric Loss used here produces a sparse heatmap of head positions. It's sparse due to the nature how the geometric loss works and the head positions are not blurred or modifyed beforehand. However, because the result is downsampled to 1/8 width and height people can still be very close together so upsampling it back to original resolution will result in local maxima. The maxpool + == trick is used to detect local maxima in an efficient way. You end up with a binary map so you multiply the original values back in to get confidence scores. Then you remove points smaller 0.05. This is not a very elegant solution but it gets the job somewhat done. The nonzeros function gives you the y,x coordinates of the remaining points that you require for localization.
